### ISSUE TITLE
out-of-line implementation of a constrained inner template of a class template specialization does not match primary template

### ISSUE THREAD

**Issue by ldalessa** [llvm/llvm-project#61414]
In the following code, Clang reports that it can not match the specialization with anything in the primary template. Both GCC and NVC++ (EDG+LLVM) accept this as valid.

```c++
#include <concepts>

template<bool>
struct a
{
    template<std::movable T>
    void c(T&& t);
};

template<>
template<std::movable T>
void a<true>::c(T&&t)
{}
```

Output:
```console
<source>:12:15: error: out-of-line definition of 'c' does not match any declaration in 'a<true>'
void a<true>::c(T&&t)
              ^
```

Live example: https://godbolt.org/z/3PnKj4c75

**Comment by shafik** [comment]
I think this should be valid but I will let @erichkeane make the call.

**Comment by usx95** [comment]
CC @alexander-shaposhnikov who has been working on similar out-of-line definition bugs.

**Comment by alexander-shaposhnikov** [comment]
A slightly simpler reproducer:
```c++
//#include <concepts>

template <class T>
concept Concept = (sizeof(T) >= 2 * sizeof(int));

template<bool>
struct a
{
  template<Concept T>
  void c(T&& t);
};

template<>
template<Concept T>
void a<true>::c(T&&t)
{}
```

```
Inside clang::Sema::IsOverload:
(lldb) p NewTemplate->dump()
FunctionTemplateDecl 0x555564be3560 parent 0x555564bc52e8 </home/co.cpp:14:1, line:15:15> col:15 c
|-TemplateTypeParmDecl 0x555564bc50e0 <line:14:10, col:18> col:18 referenced Concept 0x555564bc46f8 'Concept' depth 0 index 0 T
| `-ConceptSpecializationExpr 0x555564bc5230 <col:10> 'bool' Concept 0x555564bc46f8 'Concept'
|   |-ImplicitConceptSpecializationDecl 0x555564bc51c0 <line:4:9> col:9
|   | `-TemplateArgument type 'type-parameter-0-0'
|   |   `-TemplateTypeParmType 0x555564bc4580 'type-parameter-0-0' dependent depth 0 index 0
|   `-TemplateArgument <line:14:18> type 'T'
|     `-TemplateTypeParmType 0x555564bc5180 'T' dependent depth 0 index 0
|       `-TemplateTypeParm 0x555564bc50e0 'T'
`-CXXMethodDecl 0x555564be3460 parent 0x555564bc52e8 <line:13:1, line:15:15> col:15 c 'void (T &&)'
  `-ParmVarDecl 0x555564be3350 <col:17, col:20> col:20 t 'T &&'

(lldb) p OldTemplate->dump()
FunctionTemplateDecl 0x555564be3220 </home/co.cpp:9:3, line:10:15> col:8 c
|-TemplateTypeParmDecl 0x555564be2f20 <line:9:12, col:20> col:20 Concept 0x555564bc46f8 'Concept' depth 0 index 0 T
| `-ConceptSpecializationExpr 0x555564bc4d40 <col:12> 'bool' Concept 0x555564bc46f8 'Concept'
|   |-ImplicitConceptSpecializationDecl 0x555564bc4cd0 <line:4:9> col:9
|   | `-TemplateArgument type 'type-parameter-1-0'
|   |   `-TemplateTypeParmType 0x555564bc4c60 'type-parameter-1-0' dependent depth 1 index 0
|   `-TemplateArgument <line:9:20> type 'T'
|     `-TemplateTypeParmType 0x555564bc4c90 'T' dependent depth 1 index 0
|       `-TemplateTypeParm 0x555564bc4bb8 'T'
`-CXXMethodDecl 0x555564be3170 <line:10:3, col:15> col:8 c 'void (T &&)'
  `-ParmVarDecl 0x555564be3070 <col:10, col:14> col:14 t 'T &&'
```
I haven't fully investigated this particular case, but the differences in TemplateTypeParmType look suspicious

**Comment by erichkeane** [comment]
Yep, I would expect that to  work.  I'm guessing we're just not comparing the constraints correctly again.

**Comment by alexander-shaposhnikov** [comment]
I'm looking into this issue, hope to have an update ~soon.

**Comment by erichkeane** [comment]
> I'm looking into this issue, hope to have an update ~soon.

Ok, cool!  Let me know how you go.  At the moment, I've debugged it down to `SemaTemplate.cpp:8138` `TemplateParameterListsAreEqual` isn't returning true.  It is only hit 1x in your example, so at least that should be reasonably easy to hit in a debugger.

**Comment by erichkeane** [comment]
Grr... as soon as I posted that, I got us down quite a bit further :)  The problem is that the 'depth' of each of these template arguments are different, but we don't have the "InstantiatedFrom" for them, since they aren't being instantiated.

See SemaTemplate.cpp:8081, which checks if the constraint expressions are equal.  `AreConstraintExpressionsEqual` does its 'depth' equaling by checking where it was instantiated from.  However, in this case, we likely need to give it the old/new declaration itself (since it is, in effect, the 'instantiated from').

Let me know if that isn't clear.
